crates/sparrow-transforms/src/transforms/project.rs

```diff
@@ -0,0 +1,51 @@
+use arrow_array::RecordBatch;
+use arrow_schema::SchemaRef;
+use error_stack::{IntoReport, ResultExt};
+use sparrow_arrow::Batch;
+
+use crate::transforms::Transform;
+use crate::Error;
+use sparrow_expressions::ExpressionExecutor;
+
+/// Transform for projection.
+pub struct Project {
+    evaluators: ExpressionExecutor,
+    outputs: Vec<usize>,
+    schema: SchemaRef,
+}
+
+impl Project {
+    pub fn try_new(
+        input_schema: &SchemaRef,
+        exprs: Vec<sparrow_physical::Expr>,
+        outputs: Vec<usize>,
+        schema: SchemaRef,
+    ) -> error_stack::Result<Self, Error> {
+        let evaluators = ExpressionExecutor::try_new(input_schema.as_ref(), &exprs)
+            .change_context_lazy(|| Error::CreateTransform("project"))?;
+        Ok(Self {
+            evaluators,
+            outputs,
+            schema,
+        })
+    }
+}
+
+impl Transform for Project {
+    fn apply(&self, batch: Batch) -> error_stack::Result<Batch, Error> {
+        assert!(!batch.is_empty());
+
+        let error = || Error::ExecuteTransform("project");
+        let columns = self.evaluators.execute(&batch).change_context_lazy(error)?;
+        let columns = self
+            .outputs
+            .iter()
+            .map(|index| columns[*index].clone())
+            .collect();
```

Comment:
Does this imply that a the evaluators may produce columns that the transform doesn't need, and just drops?

Reply:
Yes. Consider: `output: a + b + c`. The only column we need to project is `output`, but to compute that we first need to compute `a + b` and then use that to compute `a + b + c`. So we'll end up with expressions like:

```
0: a
1: b
2: 0 + 1
3: c
4: 2 + 3
```

And the only output column will be column 4.

Reply:
?